File: src/tools/container_job.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { GraphQLClient } from 'graphql-request';
import { z } from 'zod';
import {
  ContainerJobListDocument,
  ContainerJobByNameDocument,
  ContainerJobCreateDocument,
  ContainerJobModifyDocument,
  ContainerJobDeleteDocument,
  type ContainerJobListQuery,
  type ContainerJobByNameQuery,
  type ContainerJobCreateMutation,
  type ContainerJobModifyMutation,
} from '../generated/graphql.js';

const EnvironmentVariable = z.object({
  name: z.string(),
  value: z.string(),
  secret: z.boolean(),
  state: z.enum(['PRESENT', 'ABSENT']),
});

const MountVolume = z.object({
  name: z.string(),
  autoCreate: z.boolean(),
  increase: z.boolean(),
  size: z.number().optional().describe('Size in GB'),
});

const Mount = z.object({
  path: z.string().describe('Mount path inside the container'),
  volume: MountVolume,
  state: z.enum(['PRESENT', 'ABSENT']),
});

export function registerContainerJobTools(server: McpServer, client: GraphQLClient): void {
  server.registerTool(
    'nexaa_container_job_list',
    {
      description: 'List all container jobs in a namespace',
      inputSchema: { namespace: z.string() },
    },
    async ({ namespace }) => {
      const data = await client.request<ContainerJobListQuery>(ContainerJobListDocument, {
        namespaceName: namespace,
      });
      return {
        content: [
          { type: 'text', text: JSON.stringify(data.namespace?.containerJobs, null, 2) },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_container_job_get',
    {
      description: 'Get a container job by name',
      inputSchema: {
        namespace: z.string(),
        name: z.string(),
      },
    },
    async ({ namespace, name }) => {
      const data = await client.request<ContainerJobByNameQuery>(ContainerJobByNameDocument, {
        namespaceName: namespace,
        containerName: name,
      });
      return {
        content: [{ type: 'text', text: JSON.stringify(data.containerJob, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_container_job_create',
    {
      description:
        "Create a new container job. schedule is a cron expression e.g. '0 * * * *'. resources e.g. CPU_250_RAM_500.",
      inputSchema: {
        namespace: z.string(),
        name: z.string(),
        image: z.string(),
        resources: z.string().describe('Resource profile e.g. CPU_250_RAM_500'),
        schedule: z.string().describe("Cron expression e.g. '0 * * * *'"),
        enabled: z.boolean().default(true),
        registry: z.string().optional(),
        command: z.array(z.string()).default([]),
        entrypoint: z.array(z.string()).default([]),
        environmentVariables: z.array(EnvironmentVariable).default([]),
        mounts: z.array(Mount).default([]),
      },
    },
    async (input) => {
      const data = await client.request<ContainerJobCreateMutation>(ContainerJobCreateDocument, {
        scheduledJob: input,
      });
      return {
        content: [{ type: 'text', text: JSON.stringify(data.containerJobCreate, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_container_job_modify',
    {
      description:
        'Modify an existing container job. namespace and name are required; all other fields are optional.',
      inputSchema: {
        namespace: z.string(),
        name: z.string(),
        image: z.string().optional(),
        resources: z.string().optional(),
        schedule: z.string().optional(),
        enabled: z.boolean().optional(),
        registry: z.string().optional(),
        command: z.array(z.string()).optional(),
        entrypoint: z.array(z.string()).optional(),
        environmentVariables: z.array(EnvironmentVariable).optional(),
        mounts: z.array(Mount).optional(),
      },
    },
    async (input) => {
      const data = await client.request<ContainerJobModifyMutation>(ContainerJobModifyDocument, {
        scheduledJob: input,
      });
      return {
        content: [{ type: 'text', text: JSON.stringify(data.containerJobModify, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_container_job_delete',
    {
      description: 'Delete a container job',
      inputSchema: {
        namespace: z.string(),
        name: z.string(),
      },
    },
    async ({ namespace, name }) => {
      await client.request(ContainerJobDeleteDocument, {
        namespaceName: namespace,
        containerJobName: name,
      });
      return {
        content: [
          {
            type: 'text',
            text: `Container job "${name}" deleted from namespace "${namespace}".`,
          },
        ],
      };
    },
  );
}

File: tools/container_job.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nexaa-cloud/nexaa-cli/api"
)

func RegisterContainerJob(s *server.MCPServer) {
	s.AddTool(
		mcp.NewTool("nexaa_container_job_list",
			mcp.WithDescription("List all container jobs in a namespace"),
			mcp.WithString("namespace", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			ns := req.GetString("namespace", "")
			result, err := withAPI(func(c *api.Client) ([]api.ContainerJobResult, error) {
				return c.ContainerJobList(ns)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_container_job_get",
			mcp.WithDescription("Get a container job by name"),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("name", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			ns := req.GetString("namespace", "")
			name := req.GetString("name", "")
			result, err := withAPI(func(c *api.Client) (api.ContainerJobResult, error) {
				return c.ContainerJobByName(ns, name)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_container_job_create",
			mcp.WithDescription("Create a new container job. schedule is a cron expression e.g. '0 * * * *'. resources e.g. CPU_250_RAM_500."),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("image", mcp.Required()),
			mcp.WithString("resources", mcp.Required(), mcp.Description("Resource profile e.g. CPU_250_RAM_500")),
			mcp.WithString("schedule", mcp.Required(), mcp.Description("Cron expression e.g. '0 * * * *'")),
			mcp.WithBoolean("enabled", mcp.Description("Enable the job"), mcp.DefaultBool(true)),
			mcp.WithString("registry"),
			mcp.WithArray("command", mcp.WithStringItems()),
			mcp.WithArray("entrypoint", mcp.WithStringItems()),
			mcp.WithArray("environmentVariables", mcp.Items(envVarItemSchema)),
			mcp.WithArray("mounts", mcp.Items(mountItemSchema)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				Namespace            string      `json:"namespace"`
				Name                 string      `json:"name"`
				Image                string      `json:"image"`
				Resources            string      `json:"resources"`
				Schedule             string      `json:"schedule"`
				Enabled              bool        `json:"enabled"`
				Registry             string      `json:"registry"`
				Command              []string    `json:"command"`
				Entrypoint           []string    `json:"entrypoint"`
				EnvironmentVariables []envVarArg `json:"environmentVariables"`
				Mounts               []mountArg  `json:"mounts"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			// mcp.DefaultBool only annotates the JSON schema; BindArguments is a
			// plain json.Unmarshal that doesn't apply it, so an omitted "enabled"
			// would otherwise leave args.Enabled at its Go zero value (false)
			// instead of the documented default of true.
			if _, ok := req.GetArguments()["enabled"]; !ok {
				args.Enabled = true
			}
			input := api.ContainerJobCreateInput{
				Namespace:            args.Namespace,
				Name:                 args.Name,
				Image:                args.Image,
				Resources:            api.ContainerResources(args.Resources),
				Schedule:             args.Schedule,
				Enabled:              args.Enabled,
				Command:              args.Command,
				Entrypoint:           args.Entrypoint,
				EnvironmentVariables: toEnvVars(args.EnvironmentVariables),
				Mounts:               toMounts(args.Mounts),
			}
			if args.Registry != "" {
				input.Registry = &args.Registry
			}
			result, err := withAPI(func(c *api.Client) (api.ContainerJobResult, error) {
				return c.ContainerJobCreate(input)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_container_job_modify",
			mcp.WithDescription("Modify an existing container job. namespace and name are required; all other fields are optional."),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("image"),
			mcp.WithString("resources"),
			mcp.WithString("schedule"),
			mcp.WithBoolean("enabled"),
			mcp.WithString("registry"),
			mcp.WithArray("command", mcp.WithStringItems()),
			mcp.WithArray("entrypoint", mcp.WithStringItems()),
			mcp.WithArray("environmentVariables", mcp.Items(envVarItemSchema)),
			mcp.WithArray("mounts", mcp.Items(mountItemSchema)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				Namespace            string      `json:"namespace"`
				Name                 string      `json:"name"`
				Image                *string     `json:"image"`
				Resources            *string     `json:"resources"`
				Schedule             *string     `json:"schedule"`
				Enabled              *bool       `json:"enabled"`
				Registry             *string     `json:"registry"`
				Command              []string    `json:"command"`
				Entrypoint           []string    `json:"entrypoint"`
				EnvironmentVariables []envVarArg `json:"environmentVariables"`
				Mounts               []mountArg  `json:"mounts"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			input := api.ContainerJobModifyInput{
				Namespace:            args.Namespace,
				Name:                 args.Name,
				Image:                args.Image,
				Registry:             args.Registry,
				Schedule:             args.Schedule,
				Enabled:              args.Enabled,
				Command:              args.Command,
				Entrypoint:           args.Entrypoint,
				EnvironmentVariables: toEnvVars(args.EnvironmentVariables),
				Mounts:               toMounts(args.Mounts),
			}
			if args.Resources != nil {
				r := api.ContainerResources(*args.Resources)
				input.Resources = &r
			}

			// api.ContainerJobModifyInput has no omitempty tags, so any field
			// left at its Go zero value here (because the caller omitted it)
			// would be sent as an explicit JSON null/[] and get reset instead of
			// left unchanged (see nexaa_container_modify in container.go for the
			// full explanation). environmentVariables/mounts are deliberately
			// NOT backfilled: EnvironmentVariableResult.Value can be redacted
			// for secret entries, so round-tripping it would silently wipe
			// stored secrets.
			provided := req.GetArguments()
			backfillFields := []string{"image", "resources", "registry", "schedule", "enabled", "command", "entrypoint"}
			needsCurrent := false
			for _, key := range backfillFields {
				if _, ok := provided[key]; !ok {
					needsCurrent = true
					break
				}
			}
			if needsCurrent {
				current, err := withAPI(func(c *api.Client) (api.ContainerJobResult, error) {
					return c.ContainerJobByName(args.Namespace, args.Name)
				})
				if err != nil {
					return mcp.NewToolResultError(err.Error()), nil
				}
				if _, ok := provided["image"]; !ok {
					input.Image = &current.Image
				}
				if _, ok := provided["resources"]; !ok {
					input.Resources = &current.Resources
				}
				if _, ok := provided["registry"]; !ok && current.PrivateRegistry != nil {
					input.Registry = &current.PrivateRegistry.Name
				}
				if _, ok := provided["schedule"]; !ok {
					input.Schedule = &current.Schedule
				}
				if _, ok := provided["enabled"]; !ok {
					input.Enabled = &current.Enabled
				}
				if _, ok := provided["command"]; !ok {
					input.Command = current.Command
				}
				if _, ok := provided["entrypoint"]; !ok {
					input.Entrypoint = current.Entrypoint
				}
			}

			result, err := withAPI(func(c *api.Client) (api.ContainerJobResult, error) {
				return c.ContainerJobModify(input)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_container_job_delete",
			mcp.WithDescription("Delete a container job. WARNING: This is irreversible. Requires confirm: true."),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("name", mcp.Required()),
			mcp.WithBoolean("confirm", mcp.Required(), mcp.Description("Must be true to confirm deletion")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			ns := req.GetString("namespace", "")
			name := req.GetString("name", "")
			if ok, cancelled := confirmDelete(req, "container job"); !ok {
				return cancelled, nil
			}
			_, err := withAPI(func(c *api.Client) (bool, error) {
				return c.ContainerJobDelete(ns, name)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("Container job %q deleted from namespace %q.", name, ns)), nil
		},
	)
}
